x/reward/keeper/query_server.go

```diff
@@ -0,0 +1,205 @@
+	case types.QueryRewardProgramsRequest_QUERY_TYPE_OUTSTANDING:
+		rewardProgramStates = []types.RewardProgram_State{types.RewardProgram_STATE_PENDING, types.RewardProgram_STATE_STARTED}
+	}
+
+	response := types.QueryRewardProgramsResponse{}
+	kvStore := sdkCtx.KVStore(k.storeKey)
+	prefixStore := prefix.NewStore(kvStore, types.RewardProgramKeyPrefix)
+	pageResponse, err := query.FilteredPaginate(prefixStore, req.Pagination, func(key []byte, value []byte, accumulate bool) (bool, error) {
+		var rewardProgram types.RewardProgram
+		vErr := rewardProgram.Unmarshal(value)
+		switch {
+		case vErr == nil && len(rewardProgramStates) == 0:
+			if accumulate {
+				response.RewardPrograms = append(response.RewardPrograms, rewardProgram)
+			}
+			return true, nil
+		case vErr == nil:
+			for _, state := range rewardProgramStates {
+				if rewardProgram.GetState() == state {
+					if accumulate {
+						response.RewardPrograms = append(response.RewardPrograms, rewardProgram)
+					}
+					return true, nil
+				}
+			}
+		default:
+			return false, vErr
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, fmt.Sprintf("unable to query all reward programs: %v", err))
+	}
+	response.Pagination = pageResponse
+	return &response, nil
+}
+
+func (k Keeper) RewardProgramByID(ctx context.Context, req *types.QueryRewardProgramByIDRequest) (*types.QueryRewardProgramByIDResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	rewardProgram, err := k.GetRewardProgram(sdkCtx, req.GetId())
+	if err != nil {
+		return &types.QueryRewardProgramByIDResponse{}, status.Errorf(codes.Internal, fmt.Sprintf("unable to query for reward program by ID: %v", err))
+	}
+	return &types.QueryRewardProgramByIDResponse{RewardProgram: &rewardProgram}, nil
+}
+
+// returns paginated ClaimPeriodRewardDistributions
+func (k Keeper) ClaimPeriodRewardDistributions(ctx context.Context, req *types.QueryClaimPeriodRewardDistributionsRequest) (*types.QueryClaimPeriodRewardDistributionsResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	response := types.QueryClaimPeriodRewardDistributionsResponse{}
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	kvStore := sdkCtx.KVStore(k.storeKey)
+	prefixStore := prefix.NewStore(kvStore, types.ClaimPeriodRewardDistributionKeyPrefix)
+	pageRes, err := query.FilteredPaginate(prefixStore, req.Pagination, func(key []byte, value []byte, accumulate bool) (bool, error) {
+		var claimPeriodRewardDist types.ClaimPeriodRewardDistribution
+		vErr := claimPeriodRewardDist.Unmarshal(value)
+		if accumulate {
+			if vErr == nil {
+				response.ClaimPeriodRewardDistributions = append(response.ClaimPeriodRewardDistributions, claimPeriodRewardDist)
+			} else {
+				return false, vErr
+			}
+		}
+		return true, nil
```

Comment:
```suggestion
		if vErr != nil {
			return false, vErr
		}
		if accumulate {
			response.ClaimPeriodRewardDistributions = append(response.ClaimPeriodRewardDistributions, claimPeriodRewardDist)
		}
		return true, nil
```

Reply:
It should be changed to something similar to this.